### PR TITLE
overlord/snapstate: increase robustness of doLinkSnap/undoLinkSnap with sanity unit tests

### PR DESCRIPTION
**Pull request by chipaca**

### CODE REVIEW

**Comment by chipaca** [comment]
DO NOT MERGE until tests are green. This fails to pass unit tests; waiting on feedback.

File: overlord/snapstate/snapmgr.go [hunk 1]
```diff
@@ -564,7 +564,17 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	}
 
 	st.Unlock()
+	// XXX: this block is slightly ugly, find a pattern when we have more examples
 	err = m.backend.LinkSnap(newInfo)
+	if err != nil {
```
Comment (niemeyer) [inline overlord/snapstate/snapmgr.go]: I may be missing something, but the logic here doesn't look right. It's undoing in-place, logging an error, and then proceeding as if everything was fine. I haven't checked the tests, but they suspect they might not be good enough too if they're accepting this situation.

Let's please talk online about this.


I'd like to talk about it live when you have a moment.

Comment (pedronis) [inline overlord/snapstate/snapmgr.go]: no, we actually return the original LinkSnap error (as we did before, indeed there are tests that cover that occurence and notice that now we are trying to remove the partial work of LinkSnap)

**Comment by pedronis** [comment]
Samuele Pedroni:
yes
 @niemeyer  the code is a bit odd (because of the locks) but it's actually checking the errors twice of LinkSnap, not just continuing as if nothing happened
 maybe there's a cleaner way to write that


[7:39:30 PM] Gustavo Niemeyer:
@pedronis So why are we undoing the snap linking in place?  Do we not have an earlier step that unliks the current one?
 @pedronis Or is that to clean up partial left overs from link snap because it's not atomic?

[7:40:11 PM] Samuele Pedroni:
@niemeyer the latter

[7:40:25 PM] Gustavo Niemeyer:
@pedronis So what happens if snapd crashes right after LinkSnap?
 @pedronis Right after the partial erroneous state is put in place

[7:40:43 PM] Samuele Pedroni:
we'll try again the whole thing
 which may succeed or fail again

[7:41:10 PM] Gustavo Niemeyer:
@pedronis Is LinkSnap accounting for that?

[7:41:18 PM] Samuele Pedroni:
yes, it can be run again and again
 that was tested in the previous branch that already landed

[7:41:36 PM] Gustavo Niemeyer:
@pedronis Okay, thanks, and sorry for the misleading comment there
 let's get this in

**Comment by pedronis** [comment]
thanks @chipaca

**Pull request merged**